FBKVOControllerTests/NSObject+FBKVOControllerTests.m

```diff
@@ -0,0 +1,59 @@
+/**
+ Copyright (c) 2014-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <XCTest/XCTest.h>
+
+#import <FBKVOController/NSObject+FBKVOController.h>
+
+#import "FBKVOTesting.h"
+
+@interface NSObject_FBKVOControllerTests : XCTestCase
+@end
+
+@implementation NSObject_FBKVOControllerTests
+
+- (void)setUp
+{
+  [super setUp];
+  FBKVOTestCircleDeallocatingWasDeallocated = NO;
+}
+
+- (void)testFBKVOControllerOnAnyObjectIsNotNil
+{
+  // Arrange: Begin observing an object using a non-retaining observer.
+  NSObject *observer = [NSObject new];
+  FBKVOTestCircleDeallocating *deallocating = [FBKVOTestCircleDeallocating circle];
+  [observer.KVOController observe:deallocating keyPath:@"radius" options:0 block:^(id observer, id object, NSDictionary *change) {
+    // noop
+  }];
+
+  // Act: Decrement the reference count of the object we're observing.
+  deallocating = nil;
+
+  XCTAssertFalse(FBKVOTestCircleDeallocatingWasDeallocated,
+                 @"The object should not have been deallocated because its observer was retaining.");
+}
+
+- (void)testFBKVOControllerNotRetainingOnAnyObjectIsNotNil
+{
+  // Arrange: Begin observing an object using a non-retaining observer.
+  NSObject *observer = [NSObject new];
+  FBKVOTestCircleDeallocating *deallocating = [FBKVOTestCircleDeallocating circle];
+  [observer.KVOControllerNonRetaining observe:deallocating keyPath:@"radius" options:0 block:^(id observer, id object, NSDictionary *change) {
+    // noop
+  }];
+
+  // Act: Release the object we're observing.
+  deallocating = nil;
```

Comment:
This line throws an exception:

```
file:///Users/bgesiak/GitHub/facebook/KVOController/FBKVOControllerTests/NSObject+FBKVOControllerTests.m: test failure: -[NSObject_FBKVOControllerTests testFBKVOControllerNotRetainingOnAnyObjectIsNotNil] failed: failed: caught "NSInternalInconsistencyException", "An instance 0x7ffd79e175c0 of class FBKVOTestCircleDeallocating was deallocated while key value observers were still registered with it. Current observation info: <NSKeyValueObservationInfo 0x7ffd79e1e260> (
<NSKeyValueObservance 0x7ffd79e24ec0: Observer: 0x7ffd7b143a20, Key path: radius, Options: <New: NO, Old: NO, Prior: NO> Context: 0x7ffd79e079c0, Property: 0x7ffd79e1e230>
)"
(
	0   CoreFoundation                      0x000000010fe50c65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x000000010fae7bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x000000010fe50b9d +[NSException raise:format:] + 205
	3   Foundation                          0x000000010f6c2332 NSKVODeallocate + 317
	4   libobjc.A.dylib                     0x000000010fafc28e _ZN11objc_object17sidetable_releaseEb + 236
	5   FBKVOControllerTests                0x00000001116d7077 -[NSObject_FBKVOControllerTests testFBKVOControllerNotRetainingOnAnyObjectIsNotNil] + 215
	6   CoreFoundation                      0x000000010fd46dec __invoking___ + 140
	7   CoreFoundation                      0x000000010fd46c42 -[NSInvocation invoke] + 290
	8   XCTest                              0x000000011175212a -[XCTestCase invokeTest] + 253
	9   XCTest                              0x000000011175232b -[XCTestCase performTest:] + 150
	10  XCTest                              0x000000011175bbe5 -[XCTest run] + 260
	11  XCTest                              0x000000011175103b -[XCTestSuite performTest:] + 379
	12  XCTest                              0x000000011175bbe5 -[XCTest run] + 260
	13  XCTest                              0x000000011175103b -[XCTestSuite performTest:] + 379
	14  XCTest                              0x000000011175bbe5 -[XCTest run] + 260
	15  XCTest                              0x000000011175103b -[XCTestSuite performTest:] + 379
	16  XCTest                              0x000000011175bbe5 -[XCTest run] + 260
	17  XCTest                              0x000000011174e0d9 __25-[XCTestDriver _runSuite]_block_invoke + 56
	18  XCTest                              0x0000000111758e8d -[XCTestObservationCenter _observeTestExecutionForBlock:] + 162
	19  XCTest                              0x000000011174e010 -[XCTestDriver _runSuite] + 269
	20  XCTest                              0x000000011174ea3d -[XCTestDriver _checkForTestManager] + 234
	21  XCTest                              0x000000011175ead0 +[XCTestProbe runTests:] + 182
	22  xctest                              0x000000010f5cc1a6 xctest + 4518
	23  xctest                              0x000000010f5cc419 xctest + 5145
	24  xctest                              0x000000010f5cbe23 xctest + 3619
	25  libdyld.dylib                       0x0000000111132145 start + 1
	26  ???                                 0x000000000000000a 0x0 + 10
)
```

So it seems like I need to unobserve objects before they're deallocated, otherwise I run into this exception. But if that's the case, what's the point of a "non-retaining" KVO controller?